### PR TITLE
bootstrap: add ability to set cert rotation base through configmap

### PR DESCRIPTION
**Pull request by soltysh**
KubeAPIServer operator has an option to change the base unit for cert
rotation. This is done through passing --unsupported-cert-rotation-base
flag to render command with a Go duration, eg. 1m will shorten the
default base to 1 minute. This commit exposes environment variable which
passes that value down to the operator.

/assign @sttts @mfojtik @wking 
/hold
this needs to wait for https://github.com/openshift/cluster-kube-apiserver-operator/pull/342 to land first.

### CODE REVIEW

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template [hunk 1]
```diff
@@ -98,6 +98,7 @@ then
 		--manifest-etcd-server-urls={{.EtcdCluster}} \
 		--manifest-image=${OPENSHIFT_HYPERSHIFT_IMAGE} \
 		--asset-input-dir=/assets/tls \
+		"{{.CertRotationBase}}" \
```
Comment (sttts) [inline data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: no flag name?

**Comment by abhinavdahiya** [comment]
/hold

I dont think env var is right way to set this.

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template [hunk 2]
```diff
@@ -98,7 +98,8 @@ then
 		--manifest-etcd-server-urls={{.EtcdCluster}} \
 		--manifest-image=${OPENSHIFT_HYPERSHIFT_IMAGE} \
 		--asset-input-dir=/assets/tls \
-		--asset-output-dir=/assets/kube-apiserver-bootstrap \
+		{{ if ne .CertRotationBase "" }}--unsupported-cert-rotation-base={{.CertRotationBase}} \
+{{ end }}		--asset-output-dir=/assets/kube-apiserver-bootstrap \
```
Comment (sttts) [inline data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: so pretty

**Comment by soltysh** [comment]
> I dont think env var is right way to set this.

I went with a similar way as `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE`, also we want to keep that deeply hidden. What else you suggest?

**Comment by abhinavdahiya** [comment]
> > I dont think env var is right way to set this.
> 
> I went with a similar way as `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE`, also we want to keep that deeply hidden. What else you suggest?

Maybe use operator's config

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template (hunk 2 above)
Comment (staebler) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: Would the following be more pleasant?

```
    {{ if ne .CertRotationBase "" -}}
    --unsupported-cert-rotation-base={{.CertRotationBase}} \
    {{ end -}}
    --asset-output-dir=/assets/kube-apiserver-bootstrap \
```

**Comment by soltysh** [comment]
> > > I dont think env var is right way to set this.
> > 
> > 
> > I went with a similar way as `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE`, also we want to keep that deeply hidden. What else you suggest?
> 
> Maybe use operator's config

We can't use operator's config because this has to be a hidden feature, it's not for public consumption. This is only for devs, QE and CI to be able to verify the cert rotation is working as should without waiting at least 15 days for cert rotation. When this will be a user (or rather admin) facing feature we'll make a proper operator config. For now it's intentionally hidden, but at the same time we need the ability to tell the installer we want to change it. Which in turn ends up calling the kube-apiserver-operator's render command with a hidden flag, which is responsible for generating additional configmap read by the operator and acted upon. Does that make sense?

**Comment by mfojtik** [comment]
@abhinavdahiya @wking reasoning for this is that we need a knob the QE and CI will turn on so we can exercise the certificate rotation mechanism and verify the cluster can survive it. We need to do this early in process to shorten the initial certificate length (we can't do this via operator config).

Without this we will risk customer clusters being bricked after 31 days.

/cc @smarterclayton

**Comment by sttts** [comment]
> I dont think env var is right way to set this.

This is super unsupported config. We need a solution soonish to change that setting. I don't think we care much whether it is an envvar, flags, config value or whatever. Just tell us what is the right way 😀

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template (hunk 2 above)
Comment (soltysh) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: @staebler thanks, I was hoping something like that existed but haven't got the time to dig :+1: updated

**Comment by soltysh** [comment]
I've discussed this with @abhinavdahiya on slack. The current approach is that we'll be following the path described in https://github.com/openshift/installer/blob/master/docs/user/overview.md#multiple-invocations to be able to create configmap directly and pushed that to the cluster.

**Comment by staebler** [comment]
> I've discussed this with @abhinavdahiya on slack. The current approach is that we'll be following the path described in https://github.com/openshift/installer/blob/master/docs/user/overview.md#multiple-invocations to be able to create configmap directly and pushed that to the cluster.

Do we need any code changes to support this path? We should be able to lay down the manifest for the configmap in the `manifests` directory without adding a specific Asset for it. The Common Manifests asset loads all file in the `manifests` directory.

**Comment by soltysh** [comment]
Per last comment this is not needed, actually. It should be already possible, closing.

**Pull request closed without merging**